### DISCUSSION TITLE
Native: Cesium 3D Tile Support

### DISCUSSION THREAD

**Discussion by sjg-wdw** [maplibre/maplibre#329] [Native feature proposals]
## Cesium 3D Tile Support for Native

Here's some functionality we'd like to add to MapLibre. This is one proposal of several for Native and if you'd like a little context and our strategy [there's an overview](https://www.wetdogweather.com/blog/maplibre-proposals).  This one is for Native, but there's a possible web component too, of course.

![3D Tiles Announcement](https://images.prismic.io/cesium/2015-08-10-nyc.png?auto=compress%2Cformat&w=800)

We think of them as Cesium's, but [3D Tiles](https://cesium.com/why-cesium/3d-tiles/) are a truly open spec and can be implemented by anyone, which makes them appealing for MapLibre.  They're also appealing because they're such a great test case for 3D interaction

### 3D Terrain and Gestures

Implementing our proposal for 3D Terrain first might be useful.  If not, we'd want to implement the gestures described in that proposal, particularly related to grabbing visual objects.

Beyond that, we'd need to look at doing intersection with 3D buildings.  If the developer wants to avoid floating through buildings, that's a fairly complex set of interactions. The game world has a number of toolkits to help out with this sort of thing so it's worth looking to that first.

### Good Shakedown Functionality

Other than looking pretty great, this would be an excellent test case for several of our other proposals. 3D Terrain, of course, but also the data loading architecture.

3D Tiles would be an excellent use for our general purpose data loading proposal and could make good use of threading, data fetching, and whatever Builder/Bucket functionality we can expose to developers as part of the Metal effort.

### Wither Globe?

This may or may not work well with a globe. Cesium is a globe, of course, but a very complex one.  It's not clear the MapLibre community is looking for that level of support for, say, orbital dynamics.  3D Tiles might be best implemented on the "flat" map, at least at first.

### Funding

Cesium has [grants](https://cesium.com/cesium-ecosystem-grants/) for development related to open source 3D.  It's worth asking if they'd like to pay for part of this.

**Comment by sjg-wdw** [comment]
The Cesium people are at FOSS4gna in Baltimore this year so I dropped by to see what they thought of this.  
They're excited at the idea.  As they explain it, the goal is to expand the ecosystem for 3D Tiles and MapLibre has a decent user base to do exactly that.

Tying a few threads together, I can see a path where you do this on Native, cross-compile it via WASM to the web and then knit it into the JS renderer.  That's.... a lot of unknowns, but an interesting approach and, as I keep suggesting, a compelling piece of functionality that the JS users would want.

In any case, big feels for 3D Tile support in MapLibre from the Cesium folks.

**Comment by HarelM** [comment]
For me, I think the main gap is not just in supporting the format in the renderer, it's also in the entire toolchain.
If I'm taking OSM data, which I think is the main source when it comes to open source, open data and MapLibre, and I need to convert it to 3D tiles, but I don't have the right tooling that can do that in a reasonable time, adding rendering support will not help with enough with the adoption, IMHO.
There has to be tools like openmaptiles, planetiler, osm2pgsql etc...
When searching this in google I get the following results, which doesn't make me want to go that path:
First results: https://groups.google.com/g/cesium-dev/c/N_P1fmZmSj8?pli=1
Second result (a pdf file?!?): https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=&cad=rja&uact=8&ved=2ahUKEwiv1vnf_o2CAxW5SKQEHdJAD5IQFnoECAwQAQ&url=http%3A%2F%2Fosgl.ethz.ch%2Ftraining%2FTutorial_FME_Cesium_3D_Tiles.pdf&usg=AOvVaw3mt7o6_PyuKWh2NKJ7H-xH&opi=89978449

I think 3D tiles are great, but there's a need to build an echo system around it, or make it more public at least.

**Reply by sjg-wdw** [reply]
There's an interesting point you're making and I think it's worth digging into it.  The Cesium and MapLibre communities really don't overlap much at all.  To a degree I find fascinating.

That's even more true of mobile map display.  I was always looking for an excuse to implement some of their data formats for our own toolkit, even before 3D Tiles and never got any takers.  It was just different markets.  I only knew about them because I have a history with the founder.

Cesium is an ecosystem and a, thus far, successful company with a strong open source bent.  That may be a community worth tapping into if it can be.  There are lots of Cesium users who could potentially be MapLibre users under the right conditions.

In terms of the tool chain, that may well be.  I haven't tried to do that myself.  I did find a [repo of tools](https://github.com/CesiumGS/3d-tiles-tools), but I haven't tried them.  FME is also a bit of a gold standard for commercial tools, but that's commercial tools.  OSM has this massive history of open data translation, so that may be a high bar to judge other projects by.

I'd also point to the "Good Shakedown Functionality" section above.  Implementation is so much easier when you have a service you can hit that returns the data you want to display.  If you do it right, you can implement the same thing with another format, which I hope is true of many of these proposals.  If MapLibre can branch out beyond GeoJSON, image and vector tile pyramids, that speaks well for the future.

But *if* MapLibre was interested in doing 3D building (and similar) display I'm not sure what else you'd pick other than 3D Tiles.  It *is* an OGC standard.  Sure, they're not all winners, but to truly understand how bad an OGC standard is, you have to implement it.

**Reply by HarelM** [reply]
I couldn't find this repo, so that's says something I guess. Also I couldn't find there a mention of geojson or OSM data.
I'm not saying every tool should support it, but if you want to try out a tool, these are the easiest to find out there and see what you get back.
Also there's no documentation site (not even using github.io) and the starts are very low - when looking into that, for me, regarding technology stack selection, this is alarming.
But might just be me :-)
In theory, if you have a backend tile server, and the rendering library support proper 3D (which currently it does not) you should be OK by converting one format to the other, like it does now in most cases where you input geojson files and you get back MVT tiles which you can present.
I'm not familiar with a better format for 3D data besides geojson with 3D points, but that's very primitive I believe and doesn't really cover full 3D.
So generally speaking, support 3D tiles in maplibre would be appreciated I believe, it will require to define a different "core" format so that  you can convert everything to it in order to present, or by other means, IDK... I also don't know how you would style a 3D map given the current style spec...

Interesting stuff :-)

**Comment by sjg-wdw** [comment]
Their main repo is [CesiumJS](https://github.com/CesiumGS/cesium) and the stats are impressive.  But I think I get it now.

Cesium predates much of the modern open source infrastructure.  Their documentation, blog, and everything else but the code is on [cesium.com](https://cesium.com/).  If you poke through old commits I bet you'll find the one where they first moved to github, rather than having started the project there.

They also don't orbit OSM.  Street maps aren't really their thing and they don't even read vector tiles or do styling in the same way.  In many ways it's a more traditional 3D approach and they have stronger ties to the gaming world.  In fact, there are Cesium versions in things like Unity.

It's just a completely different community with different priorities and ways of doing things.  But then, that's the opportunity for MapLibre.
